src/common/viz/FigureExtractor.js

```diff
@@ -17,210 +17,334 @@ define(['./Utils'], function (Utils) {
+            desc = {
+                id: id,
+                execId: execId,
+                subgraphId: nodeInfo.parent_id,
+                marker: nodeInfo.attributes.marker,
+                name: nodeInfo.attributes.name,
+                type: 'scatterPoints',
+                points: points,
+                width: nodeInfo.attributes.width,
+                color: nodeInfo.attributes.color
+            };
+
+            return desc;
+        }
+
+        getExecutionId (/* node */) {
+            throw new Error('getExecutionId is not implemented');
+        }
+
+        getGraphNode (/* node */) {
+            throw new Error('getGraphNode is not implemented');
+        }
+
+        _getContainmentParentNodeAt (/* node, metaType */) {
+            throw new Error('_getContainmentParentNodeAt is not implemented');
+        }
+
+        getMetaType (/* node */) {
+            throw new Error('getMetaType is not implmented');
+        }
+
+        GMENodeToMetadataJSON (/* node, shallow=false */) {
+            throw new Error('GMENodeTOMetadataJSON is not implemented');
+        }
+    }
+
+    class ClientFigureExtractor extends AbstractFigureExtractor {
+        constructor(client) {
+            super();
+            this._client = client;
+        }
+
+        getExecutionId (node) {
+            const executionNode = this._getContainmentParentNodeAt(node, 'Execution');
+            if (executionNode){
+                return executionNode.getId();
+            }
+        }
+
+        getMetadataChildrenIds (node) {
+            const allMetaNodes = this._client.getAllMetaNodes();
+            const metadataBaseNode = allMetaNodes
+                .find(node => node.getAttribute('name') === BASE_METADATA_TYPE);
+
+            if(metadataBaseNode) {
+                return node.getChildrenIds().filter(id => {
+                    return this._client.isTypeOf(id, metadataBaseNode.getId());
+                });
+            } else {
+                return [];
+            }
+        }
+
+        getGraphNode (node) {
+            return this._getContainmentParentNodeAt(node, 'Graph');
+        }
+
+        _getContainmentParentNodeAt (node, metaType) {
+            let currentNode = node,
+                parentId = currentNode.getParentId();
+            const isMetaType = node => this.getMetaType(node) === metaType;
+            while (parentId !== null && !isMetaType(currentNode)) {
+                currentNode = this._client.getNode(parentId);
+                parentId = currentNode.getParentId();
+            }
+            return isMetaType(currentNode) ? currentNode : null;
+        }
+
+        getMetaType (node) {
+            const metaTypeId = node.getMetaTypeId();
+            const metaNode = this._client.getNode(metaTypeId);
+            if(metaNode) {
+                return metaNode.getAttribute('name');
+            }
+        }
+
+        GMENodeToMetadataJSON (node, shallow=false) {
```

Comment:
It might be nicer to just call this `toJSON` or something.

Reply:
Also, I expect this to be essentially a private method called by `extract`.

Reply:
This method has been renamed to `toJSON`.